### PR TITLE
[Merged by Bors] - feat: Define the dual of a fractional ideal.

### PR DESCRIPTION
**Pull request by erdOne**
Also moved `Algebra.discr_eq_discr_of_toMatrix_coeff_isIntegral` to `Mathlib/NumberTheory/NumberField/Discriminant.lean`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    x ∈ Iᵛ ↔ ∀ a ∈ I, IsIntegral A (traceForm K L x a) :=
+  forall₂_congr (fun _ _ ↦ IsIntegrallyClosed.isIntegral_iff.symm)
+
+/-- If `b` is an `A`-integral basis of `L` with discriminant `b`, then `d • a * x` is integral over
+  `A` for all `a ∈ I` and `x ∈ Iᵛ`. -/
+lemma isIntegral_discr_mul_of_mem_traceFormDualSubmodule
+    (I : Submodule B L) {ι} [DecidableEq ι] [Fintype ι]
+    (b : Basis ι K L) (hb : ∀ i, IsIntegral A (b i))
+    (a x : L) (ha : a ∈ I) (hx : x ∈ Iᵛ) :
+    IsIntegral A ((discr K b) • a * x) := by
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/DedekindDomain/Different.lean:87]:
```suggestion
lemma isIntegral_discr_mul_of_mem_traceFormDualSubmodule
    (I : Submodule B L) {ι} [DecidableEq ι] [Fintype ι]
    {b : Basis ι K L} (hb : ∀ i, IsIntegral A (b i))
    {a x : L} (ha : a ∈ I) (hx : x ∈ Iᵛ) :
    IsIntegral A ((discr K b) • a * x) := by
```
Maybe the name can be changed to allow dot  notation with `Basis`, but I am not sure.

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    · rw [← (IsIntegralClosure.algebraMap_injective B A L).ne_iff, hy, RingHom.map_zero,
+        ← (algebraMap K L).map_zero, (algebraMap K L).injective.ne_iff]
+      exact discr_not_zero_of_basis K b
+    · convert isIntegral_discr_mul_of_mem_traceFormDualSubmodule I b hb _ z hx' hz using 1
+      · ext w; exact (IsIntegralClosure.isIntegral_iff (A := B)).symm
+      · rw [smul_def, RingHom.map_mul, hy, ← smul_def]⟩
+
+variable {A K}
+
+lemma FractionalIdeal.coe_dual (I : FractionalIdeal B⁰ L) (hI : I ≠ 0) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/DedekindDomain/Different.lean:165]:
```suggestion
lemma FractionalIdeal.coe_dual {I : FractionalIdeal B⁰ L} (hI : I ≠ 0) :
```

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+  · exact (smul_def _ _).symm
+
+@[simp]
+lemma FractionalIdeal.dual_eq_zero_iff {I : FractionalIdeal B⁰ L} :
+    dual A K I = 0 ↔ I = 0 := ⟨not_imp_not.mp dual_ne_zero, fun e ↦ e.symm ▸ dual_zero⟩
+
+lemma FractionalIdeal.dual_ne_zero_iff {I : FractionalIdeal B⁰ L} :
+    dual A K I ≠ 0 ↔ I ≠ 0 := dual_eq_zero_iff.not
+
+lemma FractionalIdeal.le_dual_inv_aux (I J : FractionalIdeal B⁰ L) (hI : I ≠ 0) (hIJ : I * J ≤ 1) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/DedekindDomain/Different.lean:207]:
```suggestion
lemma FractionalIdeal.le_dual_inv_aux {I J : FractionalIdeal B⁰ L} (hI : I ≠ 0) (hIJ : I * J ≤ 1) :
```

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+    1 ≤ dual A K (1 : FractionalIdeal B⁰ L) :=
+  FractionalIdeal.le_dual_inv_aux _ _ one_ne_zero (by rw [one_mul])
+
+lemma one_le_traceFormDualSubmodule_one :
+    (1 : Submodule B L) ≤ 1ᵛ := by
+  rw [← FractionalIdeal.coe_one, ← FractionalIdeal.coe_dual, FractionalIdeal.coe_le_coe]
+  exact FractionalIdeal.one_le_dual_one
+  exact one_ne_zero
+
+lemma FractionalIdeal.le_dual_iff (I J : FractionalIdeal B⁰ L) (hJ : J ≠ 0) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/DedekindDomain/Different.lean:228]:
```suggestion
lemma FractionalIdeal.le_dual_iff {I J : FractionalIdeal B⁰ L} (hJ : J ≠ 0) :
```

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+lemma FractionalIdeal.dual_eq_mul_inv (I : FractionalIdeal B⁰ L) :
+    dual A K I = dual A K 1 * I⁻¹ := by
+  by_cases hI : I = 0; · simp [hI]
+  apply le_antisymm
+  · suffices : dual A K I * I ≤ dual A K 1
+    · convert mul_right_mono I⁻¹ this using 1; simp only [mul_inv_cancel hI, mul_one, mul_assoc]
+    rw [← le_dual_iff _ _ hI]
+  rw [le_dual_iff _ _ hI, mul_assoc, inv_mul_cancel hI, mul_one]
+
+lemma FractionalIdeal.dual_mul_self (I : FractionalIdeal B⁰ L) (hI : I ≠ 0) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/DedekindDomain/Different.lean:257]:
```suggestion
lemma FractionalIdeal.dual_mul_self {I : FractionalIdeal B⁰ L} (hI : I ≠ 0) :
```

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    dual A K I * I = dual A K 1 := by
+  rw [dual_eq_mul_inv, mul_assoc, inv_mul_cancel hI, mul_one]
+
+lemma FractionalIdeal.self_mul_dual (I : FractionalIdeal B⁰ L) (hI : I ≠ 0) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/DedekindDomain/Different.lean:261]:
```suggestion
lemma FractionalIdeal.self_mul_dual {I : FractionalIdeal B⁰ L} (hI : I ≠ 0) :
```

**Review by riccardobrasca — COMMENTED** [review]
Can you try to generalize at least the definition of `traceFormDualSubmodule` as much as possible? For example I don't think it needs `K` to be a field (don't bother about theorems unless the exact same proof works).

Also, I think we want the general version for any (nondegenerate?) bilinear form. I think that a sensible thing to do is to create a `Bilinear` folder (the file `Mathlib.LinearAlgebra.BilinearForm` is very long and should be splitted), with this new notion, and then specialize to the trace in the `Different` file. If you do so feel free to ask my review.

Thanks!

**Comment by erdOne** [comment]
I did think about this, but `traceFormDualSubmodule` only works for B-middle linear forms (i.e. $\langle ra, b\rangle = \langle a, rb\rangle$), and there isn't a canonical choice of one (`(algebraMap A K).range`) in most cases. Even if both of these were given in as arguments, the results I use below mostly cannot be generalized so I refrained from doing so.

File: Mathlib/RingTheory/DedekindDomain/Different.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/RingTheory/DedekindDomain/Different.lean:85]: I don't think so, as `b : Basis` is implicit argument, and it is not the main focus of the lemma.

**Comment by erdOne** [comment]
I tried a bit, but it only generalizes well in sublattices. Or do you think a definition without much API is enough?

**Comment by erdOne** [comment]
Yes but I can't prove much about it. I can still add it if that is okay.

**Comment by erdOne** [comment]
This is now ready for review.

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    exact I.smul_mem _ (ha)
+  · exact isIntegral_trace (RingHom.IsIntegralElem.mul _ (hb j) (hb k))
+
+variable (A K)
+
+open scoped Classical
+
+/-- The dual of a non-zero fractional ideal is the dual of the submodule under the traceform. -/
+noncomputable
+def FractionalIdeal.dual (I : FractionalIdeal B⁰ L) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/DedekindDomain/Different.lean:123]: You can `namespace FractionalIdeal` here. Also `variable {I : FractionalIdeal B⁰ L}` is probably a good idea.

**Review by riccardobrasca — APPROVED** [review]
LGTM, thanks!

Can you split the file `Mathlib/RingTheory/DedekindDomain/Different.lean` into something like `Mathlib/RingTheory/Different.lean` and `Mathlib/RingTheory/DedekindDomain/Different.lean`? Files are cheap and maybe one day someone will want the different without importing all the stuff about dedekind domains.

File: Mathlib/RingTheory/DedekindDomain/Different.lean [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+variable [IsDomain A] [IsDomain B]
+variable [IsFractionRing A K] [IsIntegralClosure B A L] [IsFractionRing B L]
+variable [FiniteDimensional K L] [IsSeparable K L]
+
+open nonZeroDivisors IsLocalization Matrix Algebra
+
+/-- Under the AKLB setting, `Iᵛ := traceDual A K (I : Submodule B L)` is the
+`Submodule B L` such that `x ∈ Iᵛ ↔ ∀ y ∈ I, Tr(x, y) ∈ A` -/
+noncomputable
+def Submodule.traceDual (I : Submodule B L) : Submodule B L where
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/RingTheory/DedekindDomain/Different.lean:38]: I added `(⊥ : Submodule B L)ᵛ = ⊤`, but for the other one, the best I can get is `(⊤ : Submodule B L)ᵛ = if IsField A then ⊤ else ⊥` under `IsSeparable K L`.

[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/RingTheory/DedekindDomain/Different.lean:38]: Ouf, I though `IsSeparable` was enough to get a reasonable answer. As you want, if you've already done it I would say why not.

**Comment by erdOne** [comment]
The different ideal is the inverse of the dual which needs `IsDedekindDomain` to make sense.

**Comment by riccardobrasca** [comment]
> The different ideal is the inverse of the dual which needs `IsDedekindDomain` to make sense.

I was thinking about a file to define `Iᵛ`, but you're right it is probably not worth it.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**